doozerlib/runtime.py

```diff
@@ -794,6 +799,59 @@ def get_public_upstream(self, remote_git):
 
         return remote_https
 
+    def git_clone(self, remote_url, target_dir, gitargs=None, set_env=None, timeout=0):
+        gitargs = gitargs or []
+        set_env = set_env or []
+
+        if self.cache_dir:
+            git_cache_dir = os.path.join(self.cache_dir, self.user, 'git')
+            util.mkdirs(git_cache_dir)
+            normalized_url = util.convert_remote_git_to_https(remote_url)
+            # Strip special chars out of normalized url to create a human friendly, but unique filename
+            file_friendly_url = normalized_url.split('//')[-1].replace('/', '_')
+            repo_dir = os.path.join(git_cache_dir, file_friendly_url)
+            self.logger.info(f'Cache for {remote_url} going to f{repo_dir}')
+
+            if not os.path.exists(repo_dir):
+                self.logger.info(f'Initializing cache directory for git remote: {remote_url}')
+
+                # If the cache directory for this repo does not exist yet, we will create one.
+                # But we must do so carefully to minimize races with any other doozer instance
+                # running on the machine.
+                with self.get_dir_lock(repo_dir):  # also make sure we cooperate with other threads in this process.
+                    tmp_repo_dir = tempfile.mkdtemp(dir=git_cache_dir)
+                    exectools.cmd_assert(f'git init --bare {tmp_repo_dir}')
+                    with Dir(tmp_repo_dir):
+                        exectools.cmd_assert(f'git remote add origin {remote_url}')
+
+                    try:
+                        os.rename(tmp_repo_dir, repo_dir)
+                    except:
+                        # There are two categories of failure
+                        # 1. Another doozer instance already created the directory, in which case we are good to go.
+                        # 2. Something unexpected is preventing the rename.
+                        if not os.path.exists(repo_dir):
+                            # Not sure why the rename failed. Raise to user.
+                            raise
+
+            # If we get here, we have a bare repo with a remote set
+            with Dir(repo_dir):
+                # Pull content to update the cache. This should be safe for multiple doozer instances to perform.
+                self.logger.info(f'Updating cache directory for git remote: {remote_url}')
+                exectools.cmd_assert(f'git fetch --all', retries=3, set_env=set_env)
+                gitargs.extend(['--reference-if-able', repo_dir])
+
+        self.logger.info(f'Cloning to: {target_dir}')
+
+        # Perform the clone (including --reference args if cache_dir was set)
+        cmd = []
+        if timeout:
+            cmd.extend(['timeout', f'{timeout}'])
+        cmd.extend(['git', 'clone', remote_url])
```

Comment:
The clone has a few advantages in my mind (but I'm admittedly ignorant on the finer points of worktree):
1. The clone is completely independent and leaves no trace in the cache source.
2. The cache can be very old (possibly even corrupt) and the clone operation will only use what it can from cache.
3. In the linked doc, it says "Multiple checkout in general is still experimental, and the support for submodules is incomplete. It is NOT recommended to make multiple checkouts of a superproject."  I dunno what that means, but the warning has been around since git 2.6.0 (circa 2015). Are we sure this is solid?